ext/standard/html.c

```diff
@@ -752,6 +752,186 @@ static zend_result resolve_named_entity_html(const char *start, size_t length, c
 }
 /* }}} */
 
+/* {{{ is_codepoint_allowed */
+static inline bool is_codepoint_allowed(
+	unsigned int cp,			   /* The codepoint to check */
+	enum entity_charset charset,   /* Current charset */
+	int doctype,				   /* The doctype flags (ENT_HTML401, ENT_HTML5, etc.) */
+	const enc_to_uni *to_uni_table /* Mapping table if needed */
+	) {
+	// If charset is Unicode-compatible, the code point is used as-is
+	if (CHARSET_UNICODE_COMPAT(charset)) {
+		return unicode_cp_is_allowed(cp, doctype);
+	}
+	// If we have a mapping table (i.e., a non-UTF charset)
+	if (to_uni_table) {
+		map_to_unicode(cp, to_uni_table, &cp);
+		return unicode_cp_is_allowed(cp, doctype);
+	}
+
+	if (cp <= 0x7D) {
+		return unicode_cp_is_allowed(cp, doctype);
+	}
+
+	return true;
+}
+/* }}} */
+
+/* Lookup table for php_htmlspecialchars */
+typedef struct  {
+	char *entity[256];
+	uint8_t entity_len[256];
+} htmlspecialchars_lut;
+
+/* {{{ init_htmlspecialchars_lut */
+static void init_htmlspecialchars_lut(htmlspecialchars_lut *lut, const int flags, const int doctype) {
+	memset(lut, 0, sizeof(*lut));
+
+	lut->entity['&'] = "&amp;";
+	lut->entity['>'] = "&gt;";
+	lut->entity['<'] = "&lt;";
+	lut->entity_len['&'] = strlen(lut->entity['&']);
+	lut->entity_len['>'] = strlen(lut->entity['>']);
+	lut->entity_len['<'] = strlen(lut->entity['<']);
+
+	if (flags & ENT_QUOTES & ENT_HTML_QUOTE_DOUBLE) {
+		lut->entity['"'] = "&quot;";
+		lut->entity_len['"'] = strlen(lut->entity['"']);
+	}
+
+	if (flags & ENT_QUOTES & ENT_HTML_QUOTE_SINGLE) {
+		char *apos = "&#039;";
+		if (doctype != ENT_HTML401) {
+			if (doctype & (ENT_XML1 | ENT_XHTML | ENT_HTML5)) {
+				apos = "&apos;";
+			}
+		}
+		lut->entity['\''] = apos;
+		lut->entity_len['\''] = strlen(apos);
+	}
+}
+/* }}} */
+
+static unsigned int validate_utf8_char(
```

Comment:
This is basically slightly modified version of the `next_get_char`. So if there is some optimization, it might be just better to optimize it in `get_next_char` rather than duplicating to the completely new function.

Reply:
I can see few early returns that might save few instruction but not sure if it's going to make much difference but if it does, I think it's good to optimize in `get_next_char`